### PR TITLE
Evaluation during finetuning/unlearning

### PR DESCRIPTION
**Pull request by Dornavineeth**

### CODE REVIEW

File: src/data/utils.py [hunk 1]
```diff
@@ -57,7 +57,11 @@ def package_prompt_response(
     if chat_ids[-1] != tokenizer.eos_token_id:
         chat_ids += [tokenizer.eos_token_id]
 
-    assert chat_ids[: len(prompt_ids)] == prompt_ids
+    if template_config["asst_tag"] != "":  ## for llama2 model don't assert
+        assert chat_ids[: len(prompt_ids)] == prompt_ids, ValueError(
+            "Tokenization at the end of prompt is not same when tokenized with just pomrpt and prompt+response"
```
Comment (molereddy) [inline src/data/utils.py:62]: correct spelling

Comment (molereddy) [reply src/data/utils.py:62]: Say "Tokenization mismatch: prompt tokenization is not a prefix of prompt+response tokenization. Discrepancy usually arises around the last prompt index"

File: src/evals/base.py [hunk 2]
```diff
@@ -4,55 +4,72 @@
...
-        self.model.eval()
+        model.to(self.device)
+        model.eval()
+        return model
 
-    def load_metrics(self):
-        self.metrics_cfg = self.eval_cfg.metrics
-        self.metrics = get_metrics(self.metrics_cfg)
+    def load_metrics(self, metrics_cfg):
+        "Load metrics for evaluation"
```
Comment (molereddy) [inline src/evals/base.py:41]: use triple quotes

File: src/trainer/base.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    def evaluate(
+        self,
+        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
+        ignore_keys: Optional[List[str]] = None,
+        metric_key_prefix: str = "eval",
+    ) -> Dict[str, float]:
+        return super().evaluate(eval_dataset, ignore_keys, metric_key_prefix)
+
+    def _evaluate(self, trial, ignore_keys_for_eval, skip_scheduler=False):
+        """Performs Unlearning Evaluation pipeline and standard evaluation of HF Trainer"""
```
Comment (molereddy) [inline src/trainer/base.py:29]: Runs a custom evaluator and saves results before running the standard HuggingFace evaluate()

File: src/trainer/base.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                eval_args = {
+                    "output_dir": output_dir,
+                    "template_args": self.template_args,
+                    "model": self.model,
+                    "tokenizer": self.tokenizer,
+                }
+                self.evaluator.evaluate(**eval_args)
+            else:
+                logger.warning(
+                    "Unlearning Evaluation skipped as it supports only single gpu configuarations"
```
Comment (molereddy) [inline src/trainer/base.py:45]: Custom evaluator can be run by this Trainer only on a single GPU

File: src/trainer/unlearn/base.py [hunk 5]
```diff
@@ -53,27 +37,7 @@ def prediction_step(
...
-            prediction_loss_only (`bool`):
-                Whether or not to return the loss only.
-            ignore_keys (`List[str]`, *optional*):
-                A list of keys in the output of your model (if it is a dictionary) that should be ignored when
-                gathering predictions.
-
-        Return:
-            Tuple[Optional[torch.Tensor], Optional[torch.Tensor], Optional[torch.Tensor]]: A tuple with the loss,
-            logits and labels (each being optional).
+        Calls self.compute loss instead of compute
```
Comment (molereddy) [inline src/trainer/unlearn/base.py:40]: comment unclear

File: src/trainer/unlearn/base.py [hunk 6]
```diff
@@ -136,6 +100,7 @@ def prediction_step(
             else:
                 if has_labels or loss_without_labels:
                     with self.compute_loss_context_manager():
+                        ### Call compute_loss of super class since unlearning compute loss might not be applicable to evaluation dataset.
```
Comment (molereddy) [inline src/trainer/unlearn/base.py:103]: dont understand how finding super class (FinetuneTrainer)'s loss works here when the inputs are unlearning inputs, FinetuneTrainer computes loss for finetuning style inputs...

File: src/trainer/unlearn/base.py (hunk 5 above)
Comment (molereddy) [reply src/trainer/unlearn/base.py:40]: explain this comment better, and also keep the `Perform an evaluation step on `model` using `inputs`.` from the previous version, it helps understand what the function does

File: src/trainer/base.py (hunk 4 above)
Comment (molereddy) [reply src/trainer/base.py:45]: ^change warning

File: src/trainer/base.py (hunk 3 above)
Comment (molereddy) [reply src/trainer/base.py:29]: ^change comment

File: src/trainer/base.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+                checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+                output_dir = os.path.join(run_dir, checkpoint_folder, "evals")
+                os.makedirs(output_dir, exist_ok=True)
+                eval_args = {
+                    "output_dir": output_dir,
+                    "template_args": self.template_args,
+                    "model": self.model,
+                    "tokenizer": self.tokenizer,
+                }
+                self.evaluator.evaluate(**eval_args)
```
Comment (molereddy) [inline src/trainer/base.py:42]: did you check for evaluator==None at any point? i didnt find

File: src/train.py [hunk 8]
```diff
@@ -1,8 +1,12 @@
 import hydra
 from omegaconf import DictConfig
+import logging
 from data import get_data, get_collators
 from model import get_model
 from trainer import load_trainer
+from evals import get_evaluator
+
+logging.getLogger("absl").setLevel(logging.WARNING)
```
Comment (molereddy) [inline src/train.py:9]: what does this do

File: configs/trainer/GradAscent.yaml [hunk 9]
```diff
@@ -1,4 +1,7 @@
 defaults:
```
Comment (molereddy) [inline configs/trainer/GradAscent.yaml:1]: we should not import defaults from finetune.yaml into an unlearning yaml
even if values are in common better rewrite them here

File: configs/trainer/GradAscent.yaml [hunk 10]
```diff
@@ -1,4 +1,7 @@
...
   - finetune
 
-handler: GradAscent
\ No newline at end of file
+handler: GradAscent
+args:
```
Comment (molereddy) [inline configs/trainer/GradAscent.yaml:5]: we should not import defaults from finetune.yaml into an unlearning yaml
even if values are in common better rewrite them here

Comment (molereddy) [reply configs/trainer/GradAscent.yaml:5]: an unlearn config can use another unlearn config as a default but better to have all details in one place instead of searching through two parents

Comment (molereddy) [reply configs/trainer/GradAscent.yaml:5]: since warmup steps depends on the dataset size, i don't like mentioning the absolute value here
can we have 
```
warmup_steps: null
warmup_frac: 0.1
```
and do the rest of processing in the code?

Comment (molereddy) [reply configs/trainer/GradAscent.yaml:5]: never mind.. these seem like trainer arguments

File: src/trainer/unlearn/base.py (hunk 6 above)
Comment (Dornavineeth) [reply src/trainer/unlearn/base.py:103]: explin here better

**Review by molereddy — APPROVED** [review]
merge it in if the evaluator fix is done and nothing left
